File: problems/maximumDepthOfBinaryTree/python/maximum_depth_of_binary_tree.py
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    """
    Notes
    -----
    - TC: O(n)
    - SC: O(n)
    """

    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0

        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))


class SolutionDFSStack:
    """
    Notes
    -----
    - TC: O(n)
    - SC: O(n)
    """

    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0

        stack = [(root, 1)]
        max_depth = 0
        while stack:
            node, depth = stack.pop()
            max_depth = max(max_depth, depth)

            if node.right:
                stack.append((node.right, depth + 1))
            if node.left:
                stack.append((node.left, depth + 1))

        return max_depth


class SolutionBFS:
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0

        q = deque([root])
        count = 0
        while q:
            level = len(q)

            for _ in range(level):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            count += 1

        return count


if __name__ == "__main__":
    tree = TreeNode(3)
    tree.left = TreeNode(9)
    tree.right = TreeNode(20)
    tree.right.left = TreeNode(15)
    tree.right.right = TreeNode(7)

    ans = Solution()
    print(ans.maxDepth(tree))
    print(SolutionBFS().maxDepth(tree))
    print(SolutionDFSStack().maxDepth(tree))

File: problems/maximumDepthOfBinaryTree/c++/maximum_depth_of_binary_tree.cpp
//
//  Maximum_depth_of_binary_tree.cpp
//  DFS problem
//
//  Created by Wen-Ting Wang on 2018/11/14.
//

#include <iostream>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int maxDepth(TreeNode* root) {

        if(!root) return 0;
        else if(!root->left && !root->right) return 1;
        else
            return max(maxDepth(root->left), maxDepth(root->right)) + 1;
     /* redundat code
     //   if(root->left && root->right) return max(maxDepth(root->left), maxDepth(root->right)) + 1;
     //   else if(root->left) return maxDepth(root->left) + 1;
     //   else return maxDepth(root->right)+1;
      */
    }
};


//top-down
//ref http://www.voidcn.com/article/p-mquqvehe-bez.html
void Show(TreeNode *t) {
    if (!t) return;

    cout << t->val << " ";
    Show(t->left);
    Show(t->right);
}


int main(){

    TreeNode *root = new TreeNode(3);

    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);

    Show(root);
    Solution ans;

    cout <<"\n The answer: " <<  ans.maxDepth(root)<<endl;

    return 0;
}
